Feat(eos_cli_config_gen): Add ECN Support

chetryan opened the pull request:
## Change Summary

Added support for ECN configuration in uc-tx-queues and tx-queues is qos profiles.
Added support for ECN count command on the ethernet interfaces. This is required on some platforms where the count command on the qos profiles does not apply to the interfaces.

<!-- Enter short PR description -->

## Related Issue(s)

No issues raised.
- This PR will add to mainline some features that are in use by customers right now by means of custom jinja templates

## Component(s) name

`arista.avd.eos_cli_config_gen`

## Proposed changes
<!--- Describe your changes in detail -->
<!--- Describe data model implemented for new features -->

- Update the `eos_cli_config_gen` schema to support `random-detect ecn` for tx-queues and uc-tx-queues
- Update template to generate the configuration for tx-queues and uc-tx-queues
- Update templates to allow the use of tx-queue and uc-tx-queue commands within the ethernet port

## How to test
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->

- Run the molecule scenario `eos_cli_config_gen`. The `qos.cfg` should include the commands to configure ecn on tx and uc-tx queues as well as an example setting `ecn count` on an ethernet interface.
- 
## Checklist

### User Checklist

<!-- Add your own checklist using MD syntax and by replacing N/A -->
- Verify that the generated config does not have typos
- Verify that the way the schema is defined makes sense for someone using ECN for the first time
- Verify that the Documentation is clear enough.

### Repository Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

chetryan committed:
Feat(eos_cli_config_gen): Add ECN to tx queues

Added support for ECN configuration in uc-tx and tx queues
Added support for some uc-tx and tx queues commands on interfaces.

chetryan committed:
Fix(devcontainer) revert devcontainer changes

chetryan committed:
Feat(eos_config_cli_gen): Add documentation for ECN

chetryan committed:
Feat(eos_cli_config_gen): Dedicated table to ECN

chetryan committed:
Merge branch 'add_ecn' of https://github.com/chetryan/ansible-avd into add_ecn

chetryan committed:
Merge remote-tracking branch 'oriGin/devel' into add_ecn

chetryan committed:
Fix(eos_cli_config_gen): Update units for ECN

chetryan committed:
Fix(eos_cli_config_gen): Fix molecule tests

ClausHolbechArista commented on ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/qos.md:
```diff
@@ -325,6 +340,11 @@ QOS Profile: **no_qos_trust**
 | ----------- | ------------ | ----- | ---------- | ------------------ |
 | 3 | 4 | disabled | - | - |
 
+**ECN Configuration**
```
Please add some condition to avoid printing the empty table.

chetryan replied:
Added logic to skip any ECN Configuration tables if there are no tx queues or if none of them have any ECN config.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/qos_profiles.schema.yml:
```diff
@@ -78,6 +78,39 @@ keys:
+                      count:
+                        type: bool
+                        description: |
+                          Turn on random-detect ecn count
```
```suggestion
                          Enable counter for random-detect ECNs
```

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/qos_profiles.schema.yml:
```diff
@@ -78,6 +78,39 @@ keys:
+              random_detect:
+                type: dict
+                keys:
+                  ecn:
```
```suggestion
                  ecn:
                    description: Explicit Congestion Notification
```

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/qos_profiles.schema.yml:
```diff
@@ -78,6 +78,39 @@ keys:
+                          min:
+                            type: int
+                            description: |
+                              Set the random-detect ecn minimum-threshold
```
No need to repeat the key names etc so just try to help the user fill in expected values. Like for various timers we say "timeout in seconds" etc.
Same for the other keys in this area.

```suggestion
                              Minimum threshold in ???unit???
```

ClausHolbechArista replied:
After looking through the templates, I know understand that this is your own key - not part of the CLI as such. Is is a requirement that all values are using the same unit or would some users want different units?

chetryan replied:
The EOS command accepts an integer value and a unit for each of minimum-threshold and maximum-threshold. There are multiple units supported,  but I could not find any documentation whether the same command (for the same queue) can have mixed units. So here I am assuming that all values for 1 queue will be using the same unit.

chetryan replied:
I tested on a `DCS-7050DX4-32S-F` and the units used in the same command/line need to match. So I believe we have the right setup here, where only 1 unit can be specified for a single queue.

ClausHolbechArista replied:
OK, thank you for testing and confirming. Then just make the description `Minimum threshold` etc. and we keep the units like you have it.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/qos_profiles.schema.yml:
```diff
@@ -78,6 +78,39 @@ keys:
+                          units:
+                            type: str
+                            valid_values: ["segments", "bytes", "kbytes", "mbytes"]
+                          min:
```
Please look into improving the schema with `min` and `max` values if we know them for all keys in this area.
Also to catch some corner cases with YAML input and inline jinja, please add
```yaml
convert_types:
  - str
```
under all `type: int` keys.

ClausHolbechArista replied:
So because the units can be different, we cannot set a min/max. But still add the convert_types.

chetryan replied:
I set the max to the absolute maximum if the smallest unit is used with the platform that has the largest buffers. Shall I keep that or remove it altogether ? 
We know the min is 1, so I can leave it there.

ClausHolbechArista replied:
Maybe it will be better to remove the limits since it might be misleading if someone thinks they can use the max no matter which unit.

chetryan replied:
Removed the max limit but left the minimum.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/qos-profiles.j2:
```diff
@@ -61,6 +61,47 @@ QOS Profile: **{{ profile.name }}**
 {%                 endfor %}
 {%             endif %}
 {%         endif %}
+
```
So _above_ the blank line, you want to insert some logic that parse the queues looking for ecn key. Only if you find any, you include the table.

chetryan replied:
Added logic here to check if queues have any ecn thresholds defined. If there are none found, the generation of the table is skipped.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/qos-profiles.j2:
```diff
@@ -61,6 +61,47 @@ QOS Profile: **{{ profile.name }}**
+{%         if profile.uc_tx_queues is arista.avd.defined %}
+{%             for uc_tx_queue in profile.uc_tx_queues | arista.avd.natural_sort('id') %}
+{%                 set type = "Unicast" %}
+{%                 if uc_tx_queue.random_detect.ecn.threshold is arista.avd.defined %}
```
Detecting on the presence of a dict, is _not_ a good way to detect if it is enabled or not. If it is possible to just configure `random-detect ecn` with nothing else set, we should add an `enabled: <bool>` key under the dict and detect on that here.

chetryan replied:
After going thought the EOS documentation and testing on `DCS-7050DX4-32S-F` I found that `random-detect ecn` on it's own is not a valid command. I am thus marking 'min', 'max' and 'units' as required in the schema, and testing on the presence of `units` instead of using an added `enabled` bool.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/tx-queues.j2:
```diff
@@ -0,0 +1,43 @@
+{%     if tx_queue.random_detect.ecn.count is arista.avd.defined(True) %}
+      random-detect ecn count
+{%     endif %}
+{%     if tx_queue.random_detect.ecn.threshold is arista.avd.defined %}
+{%         set units = tx_queue.random_detect.ecn.threshold.units %}
+{%         if tx_queue.random_detect.ecn.threshold.min is arista.avd.defined %}
+{%             set min_threshold = " minimum-threshold " ~ tx_queue.random_detect.ecn.threshold.min ~ " " ~ units  %}
+{%         else %}
+{%             set min_threshold = "" %}
+{%         endif %}
+{%         if tx_queue.random_detect.ecn.threshold.max is arista.avd.defined %}
+{%             set max_threshold = " maximum-threshold " ~ tx_queue.random_detect.ecn.threshold.max ~ " " ~ units %}
+{%         else %}
+{%             set max_threshold = "" %}
+{%         endif %}
+{%         if tx_queue.random_detect.ecn.threshold.max_probability is arista.avd.defined %}
+{%             set max_probability = " max-mark-probability " ~ tx_queue.random_detect.ecn.threshold.max_probability %}
+{%         else %}
+{%             set max_probability = "" %}
+{%         endif %}
+{%         if tx_queue.random_detect.ecn.threshold.weight is arista.avd.defined %}
+{%             set ecn_weight = " weight " ~ tx_queue.random_detect.ecn.threshold.weight %}
+{%         else %}
+{%             set ecn_weight = "" %}
+{%         endif %}
+      random-detect ecn{{ min_threshold }}{{ max_threshold }}{{ max_probability }}{{ ecn_weight }}
+{%     endif %}
```
First see my connect on using the presence of the dict alone under the doc template.
Next, please look at other templates, on how we add CLI tokens bit by bit to _one_ variable instead of creating multiple like here.
Ex. (please adjust of course)
```j2
{% if feature.enabled is arista.avd.defined(true) %}
{%     set feature_cli = "basic command" %} <-- notice not space after the command
{%     if feature.knob1 is arista.avd.defined %}
{%         set feature_cli = feature_cli ~ " knob1 " ~ feature.knob1 %} <-- notice the leading space before the keyword, and no trailing space.
{%     endif %}
... more knobs ...
      {{ feature_cli }} <--- Just printing the CLI var which may or may not contain extra tokens.
{% endif %}
```

chetryan replied:
Thank you for pointing this out! I agree it is more elegant. Refactored both `tx-queues.j2` and `uc-tx-queues.j2`

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/qos-profiles.j2:
```diff
@@ -22,34 +22,10 @@ qos profile {{ profile.name }}
-{%         if tx_queue.shape.rate is arista.avd.defined %}
-      shape rate {{ tx_queue.shape.rate }}
-{%         endif %}
+{%         include 'eos/tx-queues.j2' %}
```
Adding these extra sub-sub templates to enable a bit of reuse will make it more complex to understand the templates later. @carlbuchmann @gmuloc where do we set the bar here? All the other features under the qos profiles could also be shared, and same for many other areas of eos_cli_config_gen.
IMO we should stay conservative on this.

No matter what we need to clearly name the files so we can identify the hierarchy. Maybe these should be `ethernet-interface-tx-queues.j2`

chetryan replied:
Happy to rename these, but wouldn't that mean that in the `qos-profiles.j2` we would be making reference to the `ethernet-interface-tx-queues.j2` ? I named it `tx-queues.j2` to make it independent of whether it was in the profile or the ethernet level.

ClausHolbechArista replied:
I understand your concern on naming, but I would rather have qos-profiles use the ethernet-interface... variant than confusing "tx-queues" with some global key. In general I would prefer not to have this reuse at all, but I understand why you do it.

chetryan replied:
Understood, renames these templates in the latest commit.

chetryan committed:
Update ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/qos_profiles.schema.yml

Co-authored-by: Claus Holbech <holbech@arista.com>

chetryan committed:
Update ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/qos_profiles.schema.yml

Co-authored-by: Claus Holbech <holbech@arista.com>

chetryan committed:
Fix(eos_cli_config_gen): template & schema fixes

Skip printing of empty ECN tables
Improve qos config generation template logic.
Add required, string conversion and max/min to schema

chetryan committed:
Feat(eos_cli_config_gen): Add global allow non-ect

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/qos_profiles.schema.yml:
```diff
@@ -78,6 +78,64 @@ keys:
+                              - str
+                            description: |
+                              Set the random-detect ecn max-mark-probability.
+                              The value should be between 1 and 100.
```
Avoid the "set the ...". Also try to write a description that is more clear text and helping the user if the key itself is not self-explanatory.
If the description is =~ the key, just remove the description.
Also the min/max are written in the docs, so you don't have to repeat in the description.
```suggestion
                              Random-detect ECN maximum mark probability.
```

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/qos-profiles.j2:
```diff
@@ -61,6 +61,52 @@ QOS Profile: **{{ profile.name }}**
 {%                 endfor %}
 {%             endif %}
 {%         endif %}
+
```
Move this blank line below the `if ns.ecn_table` to avoid double blank lines (check the molecule outputs before and after).

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/qos.j2:
```diff
@@ -10,6 +10,10 @@ QOS rewrite DSCP: **enabled**
 QOS rewrite DSCP: **disabled**
 {%     endif %}
+{%     if qos.random_detect.ecn.allow is arista.avd.defined %}
+
```
This file is a mess with blank lines (not your fault :))
Remove the blank in line 7, add one before each of the two rewrite. That way we avoid a double blank.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ethernet-interfaces.j2:
```diff
@@ -683,6 +683,16 @@ interface {{ ethernet_interface.name }}
 {%     if ethernet_interface.traffic_policy.output is arista.avd.defined %}
    traffic-policy output {{ ethernet_interface.traffic_policy.output }}
 {%     endif %}
+{%     if ethernet_interface.tx_queues is arista.avd.defined %}
```
you can remove this outer `if` since `arista.avd.natural_sort` handles it gracefully and just returns an empty list.

chetryan replied:
That's good to know - thank you! Removed the outer if.

chetryan committed:
Update ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/qos_profiles.schema.yml

Co-authored-by: Claus Holbech <holbech@arista.com>

chetryan committed:
Fix(eos_cli_conifg_gen): fix templates

Fix the documentation template, removing extra whitespace.
Rename eos templates to keep to standard.
Update the schema descriptions.

chetryan committed:
Fix(eos_cli_config_gen): Fix extra space in docs

chetryan committed:
Merge branch 'devel' into add_ecn

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/qos_profiles.schema.yml:
```diff
@@ -111,6 +165,60 @@ keys:
+                            valid_values: ["segments", "bytes", "kbytes", "mbytes"]
+                            description: |
+                              Units to be used for the threshold values.
+                              This should be one of segments, byte, kbytes, mbytes.
```
Avoid repeating the valid values.
Also should this be singular `unit` since it we set only one "unit" to be used on the other values?

chetryan replied:
Agreed

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/qos_profiles.schema.yml:
```diff
@@ -111,6 +165,60 @@ keys:
+                            convert_types:
+                              - str
+                            description: |
+                              Random-detect ecn minimum-threshold
```
Use upper case ECN in all descriptions

chetryan replied:
Agreed.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/qos.schema.yml:
```diff
@@ -31,3 +31,20 @@ keys:
+            keys:
+              allow:
+                type: str
+                valid_values: ["non-ect", "non-ect chip-based"]
```
It seems wrong to have a str value that always have to contain "non-ect".
Maybe change to dict with booleans if there are no other values?
```yaml
keys:
  allow_non_ect:
    type: dict
      keys:
        enabled:
          type: <bool>
        chip_based:
          type: <bool>
```

chetryan replied:
My doubts on this point relate to the EOS documentation, as it only mentions "non-ect chip-based" but on some platforms that I tested on, the command accepted was only "non-ect". I thought that having it as a str would allow us to easily update this for future platforms if there are differences. Happy to discuss this further.

chetryan replied:
Discussed in chat and agreed it is best to use the dictionary structure.

chetryan committed:
Refactor schema for ecn allow non-ect

chetryan committed:
Merge branch 'devel' into add_ecn

chetryan committed:
Merge branch 'devel' into add_ecn

chetryan committed:
Merge branch 'devel' into add_ecn

chetryan committed:
Merge remote-tracking branch 'upstream/devel' into add_ecn

chetryan committed:
Merge remote-tracking branch 'upstream/devel' into add_ecn

chetryan committed:
FIX: move the main.yml generated

tgodaA commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ethernet-interface-tx-queues.j2:
```diff
@@ -0,0 +1,37 @@
+      shape rate {{ tx_queue.shape.rate }}
+{% endif %}
+{% if tx_queue.random_detect is arista.avd.defined %}
+{%     if tx_queue.random_detect.ecn.count is arista.avd.defined(True) %}
```
Matching the same style

```suggestion
{%     if tx_queue.random_detect.ecn.count is arista.avd.defined(true) %}
```

tgodaA commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ethernet-interface-tx-queues.j2:
```diff
@@ -0,0 +1,37 @@
+{%     if tx_queue.random_detect.ecn.threshold.min is arista.avd.defined %}
+{%         set units = tx_queue.random_detect.ecn.threshold.units %}
+{%         set ecn_command = "random-detect ecn" %}
+{%         set ecn_command = ecn_command ~ " minimum-threshold " ~ tx_queue.random_detect.ecn.threshold.min ~ " " ~ units  %}
```
```suggestion
{%         set ecn_command = ecn_command ~ " minimum-threshold " ~ tx_queue.random_detect.ecn.threshold.min ~ " " ~ units %}
```

tgodaA commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ethernet-interface-uc-tx-queues.j2:
```diff
@@ -0,0 +1,37 @@
+{%     if uc_tx_queue.random_detect.ecn.threshold.min is arista.avd.defined %}
+{%         set units = uc_tx_queue.random_detect.ecn.threshold.units %}
+{%         set ecn_command = "random-detect ecn" %}
+{%         set ecn_command = ecn_command ~ " minimum-threshold " ~ uc_tx_queue.random_detect.ecn.threshold.min ~ " " ~ units  %}
```
```suggestion
{%         set ecn_command = ecn_command ~ " minimum-threshold " ~ uc_tx_queue.random_detect.ecn.threshold.min ~ " " ~ units %}
```

tgodaA commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ethernet-interface-uc-tx-queues.j2:
```diff
@@ -0,0 +1,37 @@
+      shape rate {{ uc_tx_queue.shape.rate }}
+{% endif %}
+{% if uc_tx_queue.random_detect is arista.avd.defined %}
+{%     if uc_tx_queue.random_detect.ecn.count is arista.avd.defined(True) %}
```
```suggestion
{%     if uc_tx_queue.random_detect.ecn.count is arista.avd.defined(true) %}
```

tgodaA commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ethernet-interfaces.j2:
```diff
@@ -676,6 +676,14 @@ interface {{ ethernet_interface.name }}
+{%         for uc_tx_queue in ethernet_interface.uc_tx_queues | arista.avd.natural_sort('id') %}
+{%             include 'eos/ethernet-interface-uc-tx-queues.j2' %}
+{%         endfor %}
+{%     endif %}
```
I believe the natural_sort checks if the var is defined or not.

```suggestion
{%     for uc_tx_queue in ethernet_interface.uc_tx_queues | arista.avd.natural_sort('id') %}
{%         include 'eos/ethernet-interface-uc-tx-queues.j2' %}
{%     endfor %}
```

carlbuchmann committed:
Apply Tony's suggestions from code review

Co-authored-by: Tony Reddy Goda <tgoda@arista.com>

carlbuchmann committed:
Merge branch 'devel' into add_ecn

tgodaA commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ethernet-interface-tx-queues.j2:
```diff
@@ -0,0 +1,37 @@
+{%         set units = tx_queue.random_detect.ecn.threshold.units %}
+{%         set ecn_command = "random-detect ecn" %}
+{%         set ecn_command = ecn_command ~ " minimum-threshold " ~ tx_queue.random_detect.ecn.threshold.min ~ " " ~ units %}
+{%         set ecn_command = ecn_command ~ " maximum-threshold " ~ tx_queue.random_detect.ecn.threshold.max ~ " " ~ units %}
```
There is no defined test added for this.

tgodaA replied:
tx_queue.random_detect.ecn.threshold.max

tgodaA replied:
Based on the documentation, as the min, max, and units are required. We should use "{%     if tx_queue.random_detect.ecn.threshold is arista.avd.defined %}"

ClausHolbechArista replied:
Good catch. Today we only create warnings on schema violations, so we have to catch it here for a while longer.

tgodaA commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ethernet-interface-uc-tx-queues.j2:
```diff
@@ -0,0 +1,37 @@
+{%         set units = uc_tx_queue.random_detect.ecn.threshold.units %}
+{%         set ecn_command = "random-detect ecn" %}
+{%         set ecn_command = ecn_command ~ " minimum-threshold " ~ uc_tx_queue.random_detect.ecn.threshold.min ~ " " ~ units %}
+{%         set ecn_command = ecn_command ~ " maximum-threshold " ~ uc_tx_queue.random_detect.ecn.threshold.max ~ " " ~ units %}
```
No defined test added for **max**

tgodaA replied:
same comment as tx-queues

chetryan replied:
Agreed, and added a test for the max too. Should be fixe in the latest commit.

tgodaA commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ethernet-interface-tx-queues.j2:
```diff
@@ -0,0 +1,37 @@
+{# eos - tx-queue profiles #}
+   !
+   tx-queue {{ tx_queue.id }}
+{% if tx_queue.comment is arista.avd.defined %}
+{%     for comment_line in tx_queue.comment.splitlines() | arista.avd.default([]) %}
+      !! {{ comment_line }}
+{%     endfor %}
+{% endif %}
+{% if tx_queue.bandwidth_percent is arista.avd.defined %}
+      bandwidth percent {{ tx_queue.bandwidth_percent }}
+{% elif tx_queue.bandwidth_guaranteed_percent is arista.avd.defined %}
+      bandwidth guaranteed percent {{ tx_queue.bandwidth_guaranteed_percent }}
+{% endif %}
+{% if tx_queue.priority is arista.avd.defined %}
+      {{ tx_queue.priority }}
+{% endif %}
+{% if tx_queue.shape.rate is arista.avd.defined %}
+      shape rate {{ tx_queue.shape.rate }}
```
I don't see schema for this. Let me know if it's there in other location.

ClausHolbechArista replied:
It is covered in the qos_profiles schema, but maybe not in ethernet-interfaces schema.

tgodaA commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ethernet-interface-tx-queues.j2:
```diff
@@ -0,0 +1,37 @@
+{% elif tx_queue.bandwidth_guaranteed_percent is arista.avd.defined %}
+      bandwidth guaranteed percent {{ tx_queue.bandwidth_guaranteed_percent }}
+{% endif %}
+{% if tx_queue.priority is arista.avd.defined %}
+      {{ tx_queue.priority }}
```
Just a string?

Why not **priority {{ tx_queue.priority }}**. I am not sure about command line.

tgodaA replied:
@chetryan

ClausHolbechArista replied:
The schema says:
```yaml
              priority:
                type: str
                valid_values: ["priority strict", "no priority"]
```
so the template is correct. Also this is not new. It was just moved from the old templates.

chetryan replied:
As Claus mentioned, I am keeping this as it exists at the moment for backward compatibility.

tgodaA replied:
Agreed. I don't have idea of the command line. So I commented.

carlbuchmann committed:
Merge branch 'devel' into add_ecn

carlbuchmann committed:
Merge branch 'devel' into add_ecn

chetryan committed:
FIX: Add test for max-threshold in j2

chetryan committed:
Merge branch 'devel' into add_ecn

chetryan committed:
FIX: Fix the issues raised in pre-commit

chetryan committed:
Merge branch 'devel' into add_ecn

gmuloc committed:
Doc: Run schema generation with libyaml

gmuloc committed:
Test: Run molecule

carlbuchmann commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/ethernet_interfaces.schema.yml:
```diff
@@ -979,6 +979,149 @@ keys:
+                              - str
+                            description: |
+                              Set the random-detect ecn max-mark-probability.
+                              The value should be between 1 and 100.
```
```suggestion
```
This information is already contained in min,max and doesn't need to be repeated. This should be corrected through the proposed schema. Also for valid values etc.

chetryan replied:
Agreed - removing such references in descriptions.

bjmeuer commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/eos_cli_config_gen.schema.yml:
```diff
@@ -2269,6 +2269,176 @@ keys:
+                              '
+                          weight:
+                            type: int
+                            min: 1
```
the min for the weight could be between 0 and 15, see cli:

ld002(config-s-test-qos-profile-NET-INFRA-txq-4)# random-detect ecn minimum-threshold 256 kbytes maximum-threshold 512 kbytes weight ?
  <0-15>  Value of ECN weight

bjmeuer commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/eos_cli_config_gen.schema.yml:
```diff
@@ -6778,6 +6972,70 @@ keys:
+                              '
+                          weight:
+                            type: int
+                            min: 1
```
the min for the weight could be between 0 and 15, see cli:

ld002(config-s-test-qos-profile-NET-INFRA-txq-4)# random-detect ecn minimum-threshold 256 kbytes maximum-threshold 512 kbytes weight ?
  <0-15>  Value of ECN weight

bjmeuer commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/eos_cli_config_gen.schema.yml:
```diff
@@ -6819,6 +7077,68 @@ keys:
+                              '
+                          weight:
+                            type: int
+                            min: 1
```
the min for the weight could be between 0 and 15, see cli:

ld002(config-s-test-qos-profile-NET-INFRA-txq-4)# random-detect ecn minimum-threshold 256 kbytes maximum-threshold 512 kbytes weight ?
  <0-15>  Value of ECN weight

bjmeuer commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/ethernet_interfaces.schema.yml:
```diff
@@ -979,6 +979,149 @@ keys:
+                              The value should be between 1 and 100.
+                          weight:
+                            type: int
+                            min: 1
```
the min for the weight could be between 0 and 15, see cli:

ld002(config-s-test-qos-profile-NET-INFRA-txq-4)# random-detect ecn minimum-threshold 256 kbytes maximum-threshold 512 kbytes weight ?
  <0-15>  Value of ECN weight

bjmeuer commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/ethernet_interfaces.schema.yml:
```diff
@@ -979,6 +979,149 @@ keys:
+                              The value should be between 1 and 100.
+                          weight:
+                            type: int
+                            min: 1
```
the min for the weight could be between 0 and 15, see cli:

ld002(config-s-test-qos-profile-NET-INFRA-txq-4)# random-detect ecn minimum-threshold 256 kbytes maximum-threshold 512 kbytes weight ?
  <0-15>  Value of ECN weight

bjmeuer commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/qos_profiles.schema.yml:
```diff
@@ -81,6 +81,60 @@ keys:
+                              Random-detect ECN maximum mark probability.
+                          weight:
+                            type: int
+                            min: 1
```
the min for the weight could be between 0 and 15, see cli:

ld002(config-s-test-qos-profile-NET-INFRA-txq-4)# random-detect ecn minimum-threshold 256 kbytes maximum-threshold 512 kbytes weight ?
  <0-15>  Value of ECN weight

chetryan replied:
Agreed - will be channging all references.

bjmeuer commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/qos_profiles.schema.yml:
```diff
@@ -117,6 +171,59 @@ keys:
+                              Random-detect ECN maximum mark probability.
+                          weight:
+                            type: int
+                            min: 1
```
the min for the weight could be between 0 and 15, see cli:

ld002(config-s-test-qos-profile-NET-INFRA-txq-4)# random-detect ecn minimum-threshold 256 kbytes maximum-threshold 512 kbytes weight ?
  <0-15>  Value of ECN weight

gmuloc committed:
Merge branch 'devel' into add_ecn

chetryan committed:
FIX(eos_cli_config_gen): fix ecn weight min

chetryan committed:
Merge branch 'devel' into add_ecn

ClausHolbechArista commented on ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/qos.md:
```diff
@@ -410,6 +428,17 @@ QOS Profile: **uc_mc_queues_test**
+| TX queue | Type | ECN Enabled | Min Threshold | Max Threshold | Max Mark Probability |
+| -------- | ---- | ----------- | ------------ | ------------- | -------------------- |
+| 1 | Unicast | False | -  | -  | - |
+| 2 | Unicast | False | 320 kbytes | 320 kbytes | 90 |
+| 4 | Unicast | False | 320 segments | 320 segments | - |
```
Is it correct here that ECN is not enabled?

ClausHolbechArista replied:
If this is the counter, we should change the heading to `ECN Counter Enabled` to make it clear. Also in eos_cli_config_gen we normally show `-` if the setting is unset.

chetryan replied:
This was a bug in the qos_profiles.j2. However, I agree that this column does not add a lot of information and it can be removed.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ethernet-interface-tx-queues.j2:
```diff
@@ -0,0 +1,42 @@
+      random-detect ecn count
+{%     endif %}
+{%     if tx_queue.random_detect.ecn.threshold.min is arista.avd.defined and tx_queue.random_detect.ecn.threshold.max is arista.avd.defined %}
+{%         set units = tx_queue.random_detect.ecn.threshold.units %}
```
Please add a defined check for `units` in the `if` above.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ethernet-interface-uc-tx-queues.j2:
```diff
@@ -0,0 +1,42 @@
+      random-detect ecn count
+{%     endif %}
+{%     if uc_tx_queue.random_detect.ecn.threshold.min is arista.avd.defined and uc_tx_queue.random_detect.ecn.threshold.max is arista.avd.defined %}
+{%         set units = uc_tx_queue.random_detect.ecn.threshold.units %}
```
Please add a defined check for `units` in the `if` above.

chetryan committed:
FIX: remove unrequired column in docs

chetryan committed:
Merge branch 'add_ecn' of https://github.com/chetryan/ansible-avd into add_ecn

chetryan committed:
Merge branch 'devel' into add_ecn

chetryan committed:
FEAT: Add support for millisecond unit

ClausHolbechArista commented on ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/qos.md:
```diff
@@ -360,6 +370,14 @@ QOS Profile: **test**
+**ECN Configuration**
+
+| TX queue | Type | Min Threshold | Max Threshold | Max Mark Probability |
+| -------- | ---- | ----------- | ------------ | ------------- | -------------------- |
```
Something looks off here

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/meta/main.yml:
Please bring this back.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/meta/main.yml:
This is the wrong location, so you probably dragged it by mistake from the parent folder.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/qos-profiles.j2:
```diff
@@ -69,6 +69,50 @@ QOS Profile: **{{ profile.name }}**
+
+**ECN Configuration**
+
+| TX queue | Type | Min Threshold | Max Threshold | Max Mark Probability |
+| -------- | ---- | ----------- | ------------ | ------------- | -------------------- |
```
```suggestion
| TX queue | Type | Min Threshold | Max Threshold | Max Mark Probability |
| -------- | ---- | ------------- | ------------- | -------------------- |
```

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/ethernet-interface-tx-queues.j2:
```diff
@@ -0,0 +1,42 @@
+{%     if tx_queue.random_detect.ecn.count is arista.avd.defined(true) %}
+      random-detect ecn count
+{%     endif %}
+{%     if tx_queue.random_detect.ecn.threshold.min is arista.avd.defined and tx_queue.random_detect.ecn.threshold.max is arista.avd.defined and tx_queue.random_detect.ecn.threshold.units %}
```
```suggestion
{%     if tx_queue.random_detect.ecn.threshold.min is arista.avd.defined and tx_queue.random_detect.ecn.threshold.max is arista.avd.defined and tx_queue.random_detect.ecn.threshold.units is arista.avd.defined %}
```

ClausHolbechArista committed:
Apply suggestions from code review

ClausHolbechArista commented on ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/qos.md:
```diff
@@ -410,6 +428,17 @@ QOS Profile: **uc_mc_queues_test**
 
+**ECN Configuration**
+
+| TX queue | Type | Min Threshold | Max Threshold | Max Mark Probability |
+| -------- | ---- | ----------- | ------------ | ------------- | -------------------- |
```
```suggestion
| TX queue | Type | Min Threshold | Max Threshold | Max Mark Probability |
| -------- | ---- | ------------- | ------------- | -------------------- |
```

ClausHolbechArista committed:
Update ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/qos.md

ClausHolbechArista committed:
Revert accidental move of meta file.

ClausHolbechArista committed:
Minor schema adjustments and uppercase ECN

chetryan committed:
FIX: update eos_cli_config_gen docs

ClausHolbechArista committed:
Rerun molecule

ClausHolbechArista committed:
Merge branch 'devel' into add_ecn

carlbuchmann committed:
update schema descriptions

carlbuchmann committed:
Merge branch 'devel' into add_ecn

carlbuchmann reviewed (approved):
LGTM

ClausHolbechArista committed:
Merge branch 'devel' into add_ecn
